rust_accumulator/src/lib.rs

```diff
@@ -136,12 +136,133 @@ pub extern "C" fn get_poly_commitment_g2(
 }
 
 #[cfg(test)]
-mod test_get_poly_commitments {
+mod tests {
     use super::*;
-    // in this test we will calculate the polynomial commitment a = g1^f(tau) and b = g2^f(tau)
-    // of the polynomial f = (x + 1)^5. Then do the pairing check
-    //       e(a,g2) = e(g1,b)
-    // where e is the bilinear pairing
+    use proptest::prelude::*;
+
+    /// Helper function: Schoolbook multiplication of two polynomials
+    fn direct_mul(left: &[Scalar], right: &[Scalar]) -> Vec<Scalar> {
+        let mut result = vec![Scalar::ZERO; left.len() + right.len() - 1];
+
+        for (i, l) in left.iter().enumerate() {
+            for (j, r) in right.iter().enumerate() {
+                result[i + j] += *l * *r;
+            }
+        }
+
+        result
+    }
+
+    // Strategy for generating random `Scalar`s
+    fn random_scalar() -> impl Strategy<Value = Scalar> {
+        any::<[u8; 32]>().prop_map(|b| {
+            Scalar::from_bytes_be(&b).unwrap_or(Scalar::ZERO) // Fallback to zero on failure
+        })
+    }
+
+    // Strategy for generating zero-filled `Scalar` vectors
+    fn zero_scalar() -> impl Strategy<Value = Scalar> {
+        Just(Scalar::ZERO)
+    }
+
+    // Test: Single coefficient polynomials
+    proptest! {
+        #[test]
+        fn test_fft_mul_single_coeff_polys(left in random_scalar(), right in random_scalar()) {
+            let fft_result = fft_mul(&[left], &[right]);
+            let expected = vec![left * right];
+            prop_assert_eq!(fft_result, expected, "Failed on single coefficient polynomials");
+        }
+    }
+
+    // Test: Polynomials with different lengths
+    proptest! {
+        #[test]
+        fn test_fft_mul_diff_lengths(left in prop::collection::vec(random_scalar(), 1..10),
+                                     right in prop::collection::vec(random_scalar(), 1..5)) {
+            // Perform FFT-based multiplication on the two polynomials
+            let fft_result = fft_mul(&left, &right);
+
+            // Perform direct (schoolbook) multiplication on the same polynomials
+            let direct_result = direct_mul(&left, &right);
+
+            // Expected length of the result
+            let expected_len = left.len() + right.len() - 1;
+
+            // Ensure the length is correct
+            prop_assert_eq!(fft_result.len(), expected_len, "Failed on mismatched polynomial lengths");
+
+            // Ensure the content is correct by comparing FFT result with direct multiplication result
+            prop_assert_eq!(fft_result, direct_result, "Failed on mismatched polynomials with different lengths: content mismatch");
+        }
+    }
+
+    // Test: Empty polynomials
+    proptest! {
```

Comment:
_For the record_

> @perturbing We might want to cover the case for empty polynomials in  `fft_mul`, since we get the following fail: 
> ```
> thread 'tests::test_fft_mul_empty_polys' panicked at src/lib.rs:201:5:
> Test failed: assertion failed: `(left == right)` 
>   left: `[]`, 
>  right: `[Scalar(0x0000000000000000000000000000000000000000000000000000000000000000)]`: Failed on case left = empty, right = non-empty with right.len() = 1 at src/lib.rs:209.
> minimal failing input: non_empty_vec = [
>     Scalar(0x0000000000000000000000000000000000000000000000000000000000000000),
> ]
> 	successes: 0
> 	local rejects: 0
> 	global rejects: 0
> ```

> Yeah, sounds good.
> 
> It's also a balance between safety and speed. In the code as it is now, I highly favored speed and less complexity.
> 
> But we can also make the interface safer, in that we introduce a type for a polynomial and create an interface where polynomials are always normalized and no trailing zero's etc. Then the zero is uniquely the empty list.